elide-core/src/main/java/com/yahoo/elide/security/PermissionManager.java

```diff
@@ -0,0 +1,278 @@
+                // Ignore and keep looking or queueing
+            }
+        }
+
+        // If nothing succeeded, we know nothing is queued up. We should fail out.
+        if (!anyPassing) {
+            throw new ForbiddenAccessException();
+        }
+    }
+
+    public <A extends Annotation> void checkFieldAwarePermissions(PersistentResource<?> resource,
+                                                                  ChangeSpec<?> changeSpec,
+                                                                  Class<A> annotationClass,
+                                                                  String field) {
+        EntityDictionary dictionary = resource.getDictionary();
+        // Check full object, then field
+        boolean entityFailed = false;
+        try {
+            A annotation = dictionary.getAnnotation(resource, annotationClass);
+            if (annotation != null) {
+                ExtractedChecks extracted = extractChecks(annotationClass, annotation);
+                boolean isAny = extracted.getAnyChecks().length > 0;
+                Class<? extends Check>[] checks = (isAny) ? extracted.getAnyChecks() : extracted.getAllChecks();
+                runPermissionChecks(checks, isAny, resource, changeSpec, annotationClass, true);
+            }
+        } catch (ForbiddenAccessException e) {
+            // Ignore this and continue on to checking our fields
+            entityFailed = true;
+        }
+
+        A annotation = dictionary.getAttributeOrRelationAnnotation(resource.getResourceClass(), annotationClass, field);
+        if (annotation != null) {
+            ExtractedChecks extracted = extractChecks(annotationClass, annotation);
+            boolean isAny = extracted.getAnyChecks().length > 0;
+            Class<? extends Check>[] checks = (isAny) ? extracted.getAnyChecks() : extracted.getAllChecks();
+            runPermissionChecks(checks, isAny, resource, changeSpec, annotationClass, true);
+        } else if (entityFailed) {
+            throw new ForbiddenAccessException();
+        }
+    }
+
+    public void executeCommitChecks() {
+        commitChecks.forEach(Supplier::get);
+    }
+
+    private void runPermissionChecks(Class<? extends Check>[] checks,
+                                     boolean isAny,
+                                     PersistentResource<?> resource,
+                                     ChangeSpec<?> changeSpec,
+                                     Class<? extends Annotation> annotationClass,
+                                     boolean isFieldAware) {
+        if (resource.getRequestScope().getSecurityMode() == SecurityMode.BYPASS_SECURITY) {
+            return;
+        }
+
+        BiFunction<Check, Optional<ChangeSpec<?>>, Boolean> checkFn =
+                (check, changespec) -> check.ok(resource.getRequestScope(), changespec);
+
+        boolean shouldDefer = CreatePermission.class.equals(annotationClass)
```

Comment:
It is. I have this documented in Check interface, but basically if a user specifies a commit check for a read-check, it would be automatically deferred (which breaks the semantics of our current system).

In summary, deferred checks are only valid for `Update` and `Create` operations. If they specified such a check on a `Read` or `Delete`, then we execute it inline with the caveat that the data they're looking at may be incomplete.